### PR TITLE
Create the onnxtxt serialization format

### PR DESCRIPTION
**Pull request by justinchuby**
Add support for serializing to and deserializing from the ONNX textual representation, and register it under the `onnxtxt` file extension and format. 

Users can now save a model to the textual representation with `onnx.save_model(model, "model.onnxtxt")`

Note that the textual format has known limitations, including (1) it doesn’t serialize tensor protos or initializers and (2) bugs like #5102. We will call this out in a documentation in a separate PR.

### CODE REVIEW

File: onnx/serialization.py [hunk 1]
```diff
@@ -141,7 +143,64 @@ def deserialize_proto(self, serialized: bytes | str, proto: _Proto) -> _Proto:
...
+            raise TypeError(
+                f"Parameter 'serialized' must be bytes or str, but got type: {type(serialized)}"
+            )
+        if isinstance(serialized, bytes):
+            serialized = serialized.decode(_ENCODING)
+        assert isinstance(serialized, str)
+        return google.protobuf.json_format.Parse(serialized, proto)
+
+
+class _TextualSerializer(onnx.serialization.ProtoSerializer):
```
Comment (justinchuby) [inline onnx/serialization.py:169]:
```suggestion
class _TextualSerializer(ProtoSerializer):
```

File: onnx/serialization.py [hunk 2]
```diff
@@ -165,8 +166,41 @@ def deserialize_proto(self, serialized: bytes | str, proto: _Proto) -> _Proto:
...
 
 
+class _TextualSerializer(ProtoSerializer):
+    """Serialize and deserialize the ONNX textual representation."""
+
+    supported_format = "onnxtxt"
+    file_extensions = frozenset({".onnxtxt"})
+
+    def serialize_proto(self, proto: _Proto) -> bytes:
+        text = onnx.printer.to_text(proto)
```
Comment (justinchuby) [inline onnx/serialization.py:176]:
```suggestion
        text = onnx.printer.to_text(proto)  # type: ignore[arg-type]
```

File: onnx/serialization.py [hunk 3]
```diff
@@ -165,8 +166,41 @@ def deserialize_proto(self, serialized: bytes | str, proto: _Proto) -> _Proto:
...
+        warnings.warn(
+            "The onnxtxt format is experimental. Please report any errors to the ONNX GitHub repository.",
+            stacklevel=2,
+        )
+        if not isinstance(serialized, (bytes, str)):
+            raise TypeError(
+                f"Parameter 'serialized' must be bytes or str, but got type: {type(serialized)}"
+            )
+        if isinstance(serialized, bytes):
+            serialized = serialized.decode(_ENCODING)
```
Comment (justinchuby) [inline onnx/serialization.py:189]:
```suggestion
            text = serialized.decode(_ENCODING)
        else:
            text = serialized
```

File: onnx/test/basic_test.py [hunk 4]
```diff
@@ -65,6 +68,8 @@ def test_save_and_load_model_when_input_is_file_name(self) -> None:
             model_path = os.path.join(temp_dir, "model.onnx")
             onnx.save_model(proto, model_path, format=self.format)
             loaded_proto = onnx.load_model(model_path, format=self.format)
+            print(proto)
+            print(loaded_proto)
```
Comment (jcwchen) [inline onnx/test/basic_test.py:72]: Are these prints intended?

File: onnx/serialization.py [hunk 5]
```diff
@@ -165,8 +167,43 @@ def deserialize_proto(self, serialized: bytes | str, proto: _Proto) -> _Proto:
...
+        else:
+            text = serialized
+        if isinstance(proto, onnx.ModelProto):
+            return onnx.parser.parse_model(text)  # type: ignore[return-value]
+        if isinstance(proto, onnx.GraphProto):
+            return onnx.parser.parse_graph(text)  # type: ignore[return-value]
+        if isinstance(proto, onnx.FunctionProto):
+            return onnx.parser.parse_function(text)  # type: ignore[return-value]
+        if isinstance(proto, onnx.NodeProto):
+            return onnx.parser.parse_node(text)  # type: ignore[return-value]
```
Comment (jcwchen) [inline onnx/serialization.py:200]: Just curious: why it needs `# type: ignore[return-value]` here?

File: onnx/serialization.py [hunk 6]
```diff
@@ -165,8 +167,43 @@ def deserialize_proto(self, serialized: bytes | str, proto: _Proto) -> _Proto:
...
+    supported_format = "onnxtxt"
+    file_extensions = frozenset({".onnxtxt"})
+
+    def serialize_proto(self, proto: _Proto) -> bytes:
+        text = onnx.printer.to_text(proto)  # type: ignore[arg-type]
+        return text.encode(_ENCODING)
+
+    def deserialize_proto(self, serialized: bytes | str, proto: _Proto) -> _Proto:
+        warnings.warn(
+            "The onnxtxt format is experimental. Please report any errors to the ONNX GitHub repository.",
```
Comment (jcwchen) [inline onnx/serialization.py:182]: nit: perhaps mention known limitations as warnings directly (if not many)? Or, once we have the documentation talking about known limitations, we can then provide the document link here.

File: onnx/test/basic_test.py (hunk 4 above)
Comment (justinchuby) [inline onnx/test/basic_test.py:72]:
```suggestion
```

Comment (justinchuby) [reply onnx/test/basic_test.py]: Good catch thanks

File: onnx/serialization.py (hunk 6 above)
Comment (justinchuby) [reply onnx/serialization.py:182]: I think we should just link it. For example one of the bugs is being fixed in https://github.com/onnx/onnx/issues/5102

**Review by jcwchen — APPROVED** [review]

File: onnx/serialization.py (hunk 5 above)
Comment (justinchuby) [reply onnx/serialization.py:200]: Looks like mypy doesn't understand NodeProto is a `_Proto`. I wonder if there's a better type we can use

**Comment by justinchuby** [comment]
I am going to merge this first and work on a page to document limitations

**Pull request merged**